### ISSUE TITLE
Support for React-Native specific properties

### ISSUE THREAD

**Issue by pavsidhu** [microsoft/typescript-styled-plugin#58]
There are several CSS properties that are specific to React-Native that are supported by `styled-components/native`. For example the following properties are part of React-Native but appear as errors when using this plugin:

```javascript
const Main = styled.View`
  shadow-color: black;
  shadow-opacity: 0.5;
  shadow-offset: 2px 2px;
`
```

Would React-Native specific properties be something to consider supporting?

**Comment by mjbvz** [comment]
This is now possible using the css-language-service api: https://github.com/Microsoft/vscode-css-languageservice/issues/111

This plugin can automatically contribute a list of valid react-native style properties. PRs welcome. Let me know if you have any questions about implementing this

**Comment by alvincrespo** [comment]
More than happy to help here. @mjbvz Is there a guide on how to add these properties?

**Comment by mjbvz** [comment]
Pass a list of `validProperties` to the language services, [here I believe](https://github.com/Microsoft/typescript-styled-plugin/blob/master/src/_language-service.ts#L91).

**Comment by lidonghua** [comment]
After inspecting source code of the plugin, I managed to find a working solution. Maybe it helps.

```
{
  "compilerOptions": {
    ....
    "plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": "shadow-color,shadow-opacity,shadow-offset"
        }
      }
    ]
  }
}
```

**Comment by ScreamZ** [comment]
Any news on this ?

**Comment by kallaspriit** [comment]
Workaround provided by @lidonghua kind of works, does anyone have an exhaustive list of react-native specific properties to use until these become supported?

**Comment by ScreamZ** [comment]
@kallaspriit You can add `tint-color` but I guess it's a step by step process once you get an error…

**Comment by DriesVS** [comment]
@kallaspriit so far I have:
`shadow-color,shadow-opacity,shadow-offset,padding-horizontal,padding-vertical,margin-vertical,margin-vertical,tint-color,aspect-ratio,elevation`

**Comment by sregg** [comment]
This works for me:

```
{
  "compilerOptions": {
    "plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": [
            "shadow-color",
            "shadow-opacity",
            "shadow-offset",
            "shadow-radius",
            "padding-horizontal",
            "padding-vertical",
            "margin-vertical",
            "margin-horizontal",
            "tint-color",
            "aspect-ratio",
            "elevation"
          ]
        }
      }
    ]
  }
}
```

notice `validProperties` is an actual array, not a csv string

**Comment by dimaip** [comment]
Now how do you disallow properties which are not supported by RN, e.g. `box-shadow`?

**Comment by mjbvz** [comment]
Could someone please submit a PR that documents `validProperties`

**Comment by arnaudambro** [comment]
So I have this warning too in VSCode: `Unknown property: 'shadow-radius' ts-styled-plugin(9999)` 

I set up a `tsconfig.json` file but actually I don't use TypeScript ! Anyway, this file did not change a thing for me. I also installed `stylelint` but it didn't change either, still the warning.

Any idea ?

I updated the properties for those in needs
```
{
  "compilerOptions": {
    "plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": [
            "aspect-ratio",
            "elevation",
            "margin-vertical",
            "margin-horizontal",
            "padding-horizontal",
            "padding-vertical",
            "resize-mode",
            "shadow-color",
            "shadow-opacity",
            "shadow-offset",
            "shadow-radius",
            "text-align-vertical",
            "tint-color"
          ]
        }
      }
    ]
  }
}
```

**Comment by robinmetral** [comment]
If your project isn't using Typescript, create a `jsconfig.json` instead of a `tsconfig.json`.

Sample `jsconfig.json` (at the root of your project):

```json
{
  "compilerOptions": {
    "plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": [
            "aspect-ratio",
            "elevation",
            "margin-vertical",
            "margin-horizontal",
            "padding-horizontal",
            "padding-vertical",
            "resize-mode",
            "shadow-color",
            "shadow-opacity",
            "shadow-offset",
            "shadow-radius",
            "text-align-vertical",
            "tint-color"
          ]
        }
      }
    ]
  },
  "include": ["src/**/*"]
}
```

**Comment by GollyJer** [comment]
Hey all. This issue was super helpful to fix missing `react-native` properties. Thanks! 👍

My next question...
Is there a way to say "**only** show these properties" in the auto-complete.  From time to time I find myself accidentally using web properties while in a React Native project. I'd love for the auto-complete to only show valid React Native properties.

I can put the list together but not sure if there's a place to put it. 😃
Is this possible? Thanks.

**Comment by rayonnunes** [comment]
You can also disable the warnings for unknown properties (at your own risk of losing some intellisense) adding in your `tsconfig.json`:

```javascript
"compilerOptions": {
    "plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "unknownProperties": "ignore" // "ignore" | "warning" | "error"
        }
      }
    ]
}
```

**Comment by lcustodio** [comment]
Any updates on this?
`Typescript + Styled Components + React Native` still not working

**Comment by ali-soltanii** [comment]
this is the list I'm using.
`"plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": [
            "shadow-color",
            "shadow-opacity",
            "shadow-offset",
            "shadow-radius",
            "padding-horizontal",
            "padding-vertical",
            "margin-vertical",
            "margin-horizontal",
            "tint-color",
            "aspect-ratio",
            "elevation",
            "include-font-padding",
            "text-align-vertical"
          ]
        }
      }
    ]`

**Comment by ali-soltanii** [comment]
This is the list I'm using.
```"plugins": [
      {
        "name": "typescript-styled-plugin",
        "lint": {
          "validProperties": [
            "shadow-color",
            "shadow-opacity",
            "shadow-offset",
            "shadow-radius",
            "padding-horizontal",
            "padding-vertical",
            "margin-vertical",
            "margin-horizontal",
            "tint-color",
            "aspect-ratio",
            "elevation",
            "include-font-padding",
            "text-align-vertical"
          ]
        }
      }
    ]

**Comment by pke** [comment]
Been 4 years. Are there any plans to add this to the official distribution config?

**Comment by karlhorky** [comment]
@mjbvz Ok, since I'm also having this problem with [`vscode-styled-components`](https://marketplace.visualstudio.com/items?itemName=styled-components.vscode-styled-components) not sure if I'm doing this contribution correctly, but I opened a PR over here:

- https://github.com/microsoft/vscode-css-languageservice/pull/317

**Comment by mjbvz** [comment]
Closing as this package has been deprecated in favor of the [official styled-components/typescript-styled-plugin fork]( https://github.com/styled-components/typescript-styled-plugin)

After updating to that fork, follow up in the styled-components repo if this is still an issue
